Add new rule `no-implicit-injections`

rtablada opened the pull request:
In response to our team need and the request from 1178, I've implemented a rule for `no-implicit-injections` to notify and fix deprecations related to https://deprecations.emberjs.com/v3.x/#toc_implicit-injections

Fixes https://github.com/ember-cli/eslint-plugin-ember/issues/1178

lin-ll commented on tests/lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,288 @@
+//------------------------------------------------------------------------------
```
Same here - can you add some test cases with multiple classes in one file with valid/invalid cases? And have you tested this on a codebase?

rtablada replied:
I've tested this on our codebase of over 300 components and 200+ routes and it was able to identify and autofix all instances (65 usecases). I also checked with configuration below:

```js
{
  
  rules: {
    'ember/no-implicit-injections': [
      2,
      {
        services: [
          // Ember Responsive Used to Auto Inject the media service in Components/Controllers
          { serviceName: 'media', moduleNames: ['Component', 'Controller'] },
          // Ember CLI Flash Used to Auto Inject the flashMessages service in all modules
          { serviceName: 'flashMessages' },
          { serviceName: 'store', moduleNames: ['Route', 'Controller'] },
        ],
      },
    ],
  }
}```

rtablada commented:
@lin-ll I've updated the test cases to include multiple class examples and converted the error message to use `messageId` to make the messages more informative by outputting the service name.

bmish commented on docs/rules/no-implicit-injections.md:
```diff
@@ -0,0 +1,115 @@
+
+## Related Rules
+
+- [no-unnecessary-service-injection-argument](no-unnecessary-service-injection-argument.md) omit service injection argument if possible
+- [no-implicit-service-injection-argument](no-unnecessary-service-injection-argument.md) require the service injection argument for all services (fix output for this rule follows this pattern)
```
These rules aren't really related since they're just about stylistic preferences, right?

rtablada replied:
Yeah I'll remove these, style changes should take effect with autofix as needed. It could cause a few reflows as fixers run again on the output of this rule, but that's common to happen with stuff like prettier or other stylistic rules.

bmish commented on docs/rules/no-implicit-injections.md:
```diff
@@ -0,0 +1,115 @@
+- [no-unnecessary-service-injection-argument](no-unnecessary-service-injection-argument.md) omit service injection argument if possible
+- [no-implicit-service-injection-argument](no-unnecessary-service-injection-argument.md) require the service injection argument for all services (fix output for this rule follows this pattern)
+
+## References
```
Let's include a link to the Ember Data `store` spec somewhere.

rtablada replied:
I'm not sure what you're referring to? Do you have a particular link in mind?

bmish replied:
https://api.emberjs.com/ember-data/release/classes/Store

Since you're discussing the Ember Data `store` service, I think it would make sense to have a link to it for someone wanting more context.

bmish commented on docs/rules/no-implicit-injections.md:
````diff
@@ -0,0 +1,115 @@
+```js
+module.exports = {
+  rules: {
+    'ember/no-implicit-injections': [2, {
````
For clarity let's use `error` instead of `2` for the severity level.

```suggestion
    'ember/no-implicit-injections': ['error', {
```

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,177 @@
+        type: 'object',
+        properties: {
+          services: {
+            type: 'array',
```
`minItems: 1` for arrays is helpful.

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,177 @@
+            items: {
+              type: 'object',
+              default: [{ serviceName: 'store', moduleNames: ['Route', 'Controller'] }],
+              properties: {
```
Let's also use `additionalProperties: false,` for this object.

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,177 @@
+    const classStack = new Stack();
+
+    // This rule does not apply to test files or non class modules
+    if (emberUtils.isTestFile(context.getFilename())) {
```
According to test coverage checks, this line is not tested. Can you test it? And also be sure to monitor the test coverage to ensure every line in the new rule is covered.

```
yarn test --coverage lib/rules/no-deprecated-router-transition-methods.js
```

rtablada replied:
I'm rethinking the need to not run this in tests. You could have mock services, routes, controllers, components, etc declared in tests that would still suffer from the same deprecation and we'd want to catch those.

bmish commented on tests/lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,369 @@
+  },
+});
+
+ruleTester.run('no-implicit-injections', rule, {
```
Since we're using the class stack, technically we should also be testing that the behavior is correct with nested classes. For example, that the behavior is correct inside the nested class, and after the nested class but still inside the outer class.

rtablada replied:
I'll look at other test scenarios for class stack to see how those setup tests.

rtablada replied:
This actually lead to finding some failures.

`ClassDeclaration` alone doesn't properly identify nested classes nor modules returned by functions.

This lead to a few more changes because of underlying issues in `ember/utils` commented in diff.

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,177 @@
+                serviceName: {
+                  type: 'string',
+                },
+                moduleNames: {
```
Are only specific values allowed for `moduleNames`? If so, we can use `"enum": ["Controller", "Component", ...]` to enforce the allowed values for it.

rtablada replied:
Added enum based on the module name options in the various `isEmber...` in `utils/ember.js`: https://github.com/rtablada/eslint-plugin-ember/blob/rt/no-implicit-injections/lib/utils/ember.js#L275-L317

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,177 @@
+                  type: 'string',
+                },
+                moduleNames: {
+                  type: 'array',
```
`minItems: 1` for arrays is helpful.

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,177 @@
+            type: 'array',
+            items: {
+              type: 'object',
+              default: [{ serviceName: 'store', moduleNames: ['Route', 'Controller'] }],
```
Can we use `defaultServiceConfig` here?

bmish commented on docs/rules/no-implicit-injections.md:
```diff
@@ -0,0 +1,115 @@
+- object -- containing the following properties:
+  - array -- `services` -- Array of configuration objects configuring the lint rule to check for use of implicit injected services
+    - string -- `serviceName` -- The name of the service that is implicitly injected
+    - array -- `moduleNames` -- Array of string listing the types of classes (controller, route, component, etc) to check for implicit injections. If an array is declared only those class types will be checked for implicit injection. (Defaults to checking all class files/types)
```
Regarding `moduleNames`, how critical do you think this option is?

Just to explore an alternative design, it's possible we could use ESLint [overrides](https://eslint.org/docs/latest/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns) to achieve mostly the same thing but with a much simpler version of the rule:

```js
// .eslintrc.js
module.exports = {
  overrides: [
    {
      files: ['controllers/**/*.js'],
      rules: {
        'ember/no-implicit-injections': ['error', { services: ['store'] }],
      },
    },
    {
      files: ['components/**/*.js', 'routes/**/*.js'],
      rules: {
        'ember/no-implicit-injections': ['error', { services: ['someOtherService'] }],
      },
    },
  ],
};
```

The user can use this built-in feature of ESLint to easily apply the rule to whatever type of file they prefer, with the full customizability that these glob file patterns provide.

And most of the time, the user won't even need to target specific types of Ember classes anyway, right? As in, I assume most of the time the default of "checking all class files/types" is sufficient. So we could reduce the complexity of the rule a lot if we don't need to worry about checking for specific Ember classes.

If you think `moduleNames` is still important, let us know why. I just want to make sure we don't add unnecessary complexity that can already be achieved with `overrides`, which is a mistake I've made myself before.

rtablada replied:
Overrides on files doesn't properly account for class types, pod structures, or the way that services are injected by the Ember initializers.

The `moduleNames` (maybe `moduleTypes` would be a better property name for it) allows for better identification since it uses `isEmberCoreModule` for module type identification to better know what kind of module is being checked from the base class.

rtablada replied:
For modules with more than one class (pod or just native classes) this could also become overly zealous.

Ex a team exports route and controller in a single module. Implicit uses of `media` are allowed in routes but not in controllers. This should be configurable and file based overrides will not account for this.

A more common example would be that a component declares the component *and* a native class. The native class is not subject to the implicit service injection rule.

---

I'll add test cases for both of these.

rtablada replied:
Test cases added

bmish replied:
Makes sense, thanks for the justification.

rtablada commented on lib/utils/ember.js:
```diff
@@ -222,7 +222,7 @@ function isEmberCoreModule(context, node, moduleName) {
     // "classic" class pattern
     return isClassicEmberCoreModule(node, moduleName, context.getFilename());
-  } else if (types.isClassDeclaration(node)) {
+  } else if (types.isClassDeclaration(node) || types.isClassExpression(node)) {
```
Class Expressions can be modules since functions can return classes that extend from Ember module types.

bmish replied:
Let's add a test for this change in `tests/lib/utils/ember-test.js`.

bmish commented on lib/rules/no-implicit-injections.js:
```diff
@@ -0,0 +1,195 @@
+              default: [defaultServiceConfig],
+              properties: {
+                serviceName: {
+                  type: 'string',
```
Let's also use `minLength: 1` for strings.

rtablada replied:
Updated and marked `required` for the values in these objects too.

bmish commented on docs/rules/no-implicit-injections.md:
```diff
@@ -0,0 +1,112 @@
+This lint rule will search for instances of `store` used in routes or controllers by default. If you have other services that you would like to check for uses of, the configuration can be overridden.
+
+- object -- containing the following properties:
+  - array -- `services` -- Array of configuration objects configuring the lint rule to check for use of implicit injected services
```
One observation is that [ember/no-restricted-service-injections](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-restricted-service-injections.md#configuration) accepts an unnamed, top-level array of objects for its schema, each with a `services` property.

I actually think what you have with a named, top-level array property is probably better practice since it allows us to add other top-level, named properties as needed.

But one way we could possibly improve yours is to name the top-level property `denylist` instead of `services` to avoid confusion between `services` and `serviceNames`. We could also optionally name the nested property `services` instead of `serviceNames` for consistency with the other rule, but I don't have a strong preference between these names.

rtablada replied:
Updated this to `denyList` and `service`

bmish commented on lib/utils/types.js:
```diff
@@ -130,6 +131,16 @@ function isClassDeclaration(node) {
+ * @param {Object} node The node to check.
+ * @returns {boolean} Whether or not the node is a ClassExpression.
+ */
+function isClassExpression(node) {
```
We probably want to avoid adding this helper. I'm not a fan of these trivial type check helpers and plan to make them deprecated.

rtablada replied:
I wasn't sure. These helpers are pretty trival, but I think it'd be weirder to have the call in `isEmberCoreModule` be `(types.isClassDeclaration(node) || node.type === 'ClassExpression')` or to do more major refactors to that code.

bmish replied:
I'm deprecating these helpers here: https://github.com/ember-cli/eslint-plugin-ember/pull/1720

It's fine if we're slightly inconsistent with the type checks while we're migrating away from them, but I do want avoid adding new trivial type check helpers.

rtablada commented:
@bmish I think I've addressed the feedback so far.

rtablada commented on docs/rules/no-implicit-injections.md:
```diff
@@ -0,0 +1,113 @@
+    'ember/no-implicit-injections': ['error', {
+      denyList: [
+        // Ember Responsive Used to Auto Inject the media service in Components/Controllers
+        { service: 'media', moduleNames: ['Component', 'Controller'] },
```
I prefer singular. Merging multiple arrays of service names and their respective module type filters could get really tricky to reason about.

Individual configs per service is fairly reasonable and allows for more explicit (though a bit duplicative) configuration.

bmish replied:
Fair enough.

rtablada commented on docs/rules/no-implicit-injections.md:
```diff
@@ -0,0 +1,112 @@
+
+- object -- containing the following properties:
+  - array -- `services` -- Array of configuration objects configuring the lint rule to check for use of implicit injected services
+    - string -- `serviceName` -- The name of the service that is implicitly injected
```
https://github.com/ember-cli/eslint-plugin-ember/pull/1714/files/#diff-ce4c727cc50236f1ac7ea5af2f03053469b2bff730a010ed6f0c25d29ab97319R90-R93

I added the check for rule behavior as requested and updated the md to reference that service names should be kebab cased. https://github.com/ember-cli/eslint-plugin-ember/pull/1714/files/#diff-ce4c727cc50236f1ac7ea5af2f03053469b2bff730a010ed6f0c25d29ab97319R90-R93

I'm not sure I follow what you're saying in the bullet list and I'm unclear on `isInjectedServiceProp` comment. I based things off the work in `no-assigment-of-untracked` and the patterns there since I found it to most closely resemble the state checks required to look for uses.

bmish replied:
The part about the input format for service names has been addressed.

But I still want to make sure every format of service name mentioned above is able to be detected by your rule when it checks if a service injection already exists. Can you add a test case checking for all possible formats of service injections (including nested service names) to ensure the rule is able to detect each one? You could just pass in ~10 service names to the denylist for one test case and then inject each one in a different format to ensure it's detected.

I was just suggesting `isInjectedServiceProp` since it could be helpful in regards to checking for service injections and handling the different formats, but not required if you already have a good way to do it.

rtablada replied:
The service definition types aren't allowed in the deny list. The deny list config has to be the service module path.

I did however add tests for ways service could be declared in a class and added/fixed use in legacy `.extends` modules

bmish replied:
Got it. Thanks.

Have you tested nested services yet? The service name could look like `my-scope/my-service`.

rtablada replied:
Though...

Now I'm realizing that `service` alone isn't really 100% coverage for uses of auto-injected services. 

The current config and code assumes a 1:1 mapping (👋🏽 around kebab and camelCasing) of the injected property name to the service name. But... This may not be the case: initializers can inject any property name they want:

```js
// Example of 1:1 mapping
application.inject('controller', 'flashMessages', 'service:flash-messages');

// Example of arbitrary mapping
application.inject('controller', 'somethingElse', 'service:flash-messages');
```

I think the config feels good for the 80% usecase (and TBH most teams will likely not configure at all and just look for implicit use of `store`.

We could augment this with a config `propertyName` which would default to `camelCase(config.service)`

bmish replied:
If we care about that edge case at all, we can worry about it as a follow-up. I'm ready to merge this PR as soon as I am sure nested services are handled.

rtablada replied:
The thought of `propertyName` came from the issue with nested services`. Ember AFAIK doesn't have a way of auto mapping a property to a nested service.

So if you have an autoInjection of `@service('cart/checkout')` there has to be an explicit `propertyName`. I'll add an assertion to check for this config.

bmish replied:
Okay. I would still like to see a test with the nested service name.

rtablada replied:
Updated and added the assertion mentioned above

rtablada commented:
@bmish updated, added tests to make it work in legacy `.extends` modules as well.

rtablada commented:
Alright! Got `propertyName` config working and updated docs. That actually cleared up some of the confusion and weird naming in the rule.

I think it's now at a point where it works for flexible config, works for legacy modules, AND the code for the rule is easier to read.

bmish reviewed (approved):
Thanks for your great work on this!
